### PR TITLE
[Perf] Bound DiffusionGemma sampler transient via request-tiled logits

### PR DESCRIPTION
**Pull request by guan404ming**
## Purpose

Bound the DiffusionGemma sampler's transient logit memory. It scales with `max_num_seqs` and OOMs before weights or the KV pool do.

## Implementation

- Tile `_compiled_sample_step` over request groups; the group size is auto-sized from free GPU memory (`mem_get_info`), no tuning knob. Tiling is bit-identical to a single pass.
- Compute logprobs per tile so the fp32 scaled-logits buffer frees each iteration.
- Clear `sampled` / `num_sampled` once before the loop (moved out of `_compiled_sample_step`) so a tile cannot wipe earlier tiles' writes.
- Transient: `O(NumDecode * Canvas * Vocab)` -> `O(group * Canvas * Vocab)`.

## Test

Compare peak memory, throughput, and GSM8K accuracy on H200 (`diffusiongemma-26B-A4B-it`):

- Sampler transient `23.9 -> 5.5 GiB`; admits `max_num_seqs=32` where the un-tiled path OOMs.
- GSM8K parity: `54.8%` vs `54.5%` (within run-to-run noise).

image

`group` size from mem_get_info(): budget 50% of free, ~10 fp32 copies per req. Verified on T4/L4/A10G/A100/H100/H200: peak stays 34–40% of free, never OOMs.

image

### CODE REVIEW

**Comment by guan404ming** [comment]
Hi @DarkLight1337, sorry to bother you again, I wasn't sure who to ping for review. This PR is a improvement for DLM (DiffusionGemma) inference. Would you mind taking a look or pinging a other suitable maintainer? Thanks for your time!

File: vllm/envs.py [hunk 1]
```diff
@@ -718,6 +719,12 @@ def _resolve_rust_frontend_path() -> str | None:
     # without re-splitting graph modules. This reduces overhead during model
     # loading by using reconstruct_serializable_fn_from_mega_artifact.
     "VLLM_USE_MEGA_AOT_ARTIFACT": use_mega_aot_artifact,
+    # DiffusionGemma: decode requests per sampler tile. > 0 tiles the sampler
+    # to bound its transient logit memory at high concurrency; 0 (default) runs
+    # a single pass.
+    "VLLM_DGEMMA_SAMPLE_CHUNK": lambda: int(
+        os.environ.get("VLLM_DGEMMA_SAMPLE_CHUNK", "0")
+    ),
```
Comment (Isotr0py) [inline vllm/envs.py]: I strongly feel that we shouldn't add an environment variable for a model-specific feature, especially this model is a dLLM edge case.

File: vllm/model_executor/models/diffusion_gemma.py [hunk 2]
```diff
@@ -1258,100 +1254,111 @@ def __call__(
...
-
-        # --- Logprobs: stash on convergence, return on commit ---
         slots_np = input_batch.idx_mapping_np[:num_reqs]
         is_decode_np = per_req_nlogits_np > 0
+        max_num_logprobs = self.sampling_states.max_num_logprobs(slots_np)
 
+        # Sample over the [num_decode * CL, vocab] logits. The fp32 pipeline
+        # peaks at several live copies, so VLLM_DGEMMA_SAMPLE_CHUNK > 0 tiles it
+        # over requests to bound the transient; 0 is a single pass (bit-identical).
+        chunk = envs.VLLM_DGEMMA_SAMPLE_CHUNK
```
Comment (Isotr0py) [inline vllm/model_executor/models/diffusion_gemma.py]: If the goal of chunking is just to avoid OOM from single pass, maybe you can estimate the necessary chunk based on free memory during sampling.

File: vllm/envs.py (hunk 1 above)
Comment (guan404ming) [reply vllm/envs.py]: That make sense to me. Just Removed. Tile size now auto-derived from free GPU memory, no env var. Matches `gemma4_mm.py::_encoder_chunk`

File: vllm/model_executor/models/diffusion_gemma.py (hunk 2 above)
Comment (guan404ming) [reply vllm/model_executor/models/diffusion_gemma.py]: That is a great idea. I modified to make `group` size from mem_get_info(): budget 50% of free, ~10 fp32 copies per req. Verified on T4/L4/A10G/A100/H100/H200: peak stays 34–40% of free, never OOMs.

image

**Comment by masterFoad** [comment]
Thanks for working on this. I independently hit the same DiffusionGemma sampler transient issue under A100 serving pressure.

I opened Draft PR #46155 with related validation and a different implementation shape. It includes A100 c16/o128/n500 ShareGPT evidence, materialized warmup OOM logs, logprobs fallback validation, and a generic model-reported non-KV reserve hook.

This is overlapping work, so I am not trying to duplicate review effort. If maintainers prefer, I can contribute the validation data/tests here, split the generic reserve hook separately, or rework my branch as a follow-up after this PR lands.

**Comment by masterFoad** [comment]
Hi, thanks for working on this. I have been working on a related DiffusionGemma sampler-memory path in #46155, so I ran a (A100) comparison against the request-tiling shape here.

Request tiling does reduce the materialized sampler transient when the group size is small. In my test at `CL=256`, `vocab=262144`, `num_decode=16` (`4096` sampler rows), the sampler transient over already-live logits dropped from `28,672 MiB` to `1,800 MiB` with request group size `1`.

One correctness detail to watch: tiling changed stochastic sampled token IDs in this fixed-seed check. Greedy tokens still matched the untiled path.

| request group | sampled mismatches | greedy mismatches |
|---:|---:|---:|
| 1 | 45 | 0 |
| 2 | 33 | 0 |
| 4 | 11 | 0 |
| 5 | 0 | 0 |

Test shape: `num_decode=5`, `CL=17`, `vocab=4097`.

FYI.

**Review by Isotr0py — APPROVED** [review]

**Comment by guan404ming** [comment]
Hi @Isotr0py @DarkLight1337 could you help merge this? The ci looks all green now. Thanks!

**Comment by guan404ming** [comment]
Hi @Isotr0py this one's ci looks great now again, could you help take another look, thanks!

**Comment by guan404ming** [comment]
Really thanks for your patience @Isotr0py!
